### PR TITLE
{AKS} az aks nodepool add/update/upgrade: add new parameter --node-soak-duration

### PR DESCRIPTION
**Pull request by MaxHorstmann**
### Related command

`az aks nodepool add/update/upgrade --node-soak-duration`

nodeSoakDuration was added to [2023-09-02-preview](https://github.com/Azure/azure-rest-api-specs/blob/5ae522bc106bf8609c6cb379e584aa3e0e2639f3/specification/containerservice/resource-manager/Microsoft.ContainerService/aks/preview/2023-09-02-preview/managedClusters.json#L4033-L4040)

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally? (`pip install wheel==0.30.0` required)

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
AKS

File: src/aks-preview/azext_aks_preview/custom.py [hunk 1]
```diff
@@ -1216,13 +1222,13 @@ def aks_agentpool_upgrade(cmd,
     )
     if kubernetes_version != '' and node_image_only:
         raise MutuallyExclusiveArgumentError(
-            'Conflicting flags. Upgrading the Kubernetes version will also '
-            'upgrade node image version. If you only want to upgrade the '
-            'node version please use the "--node-image-only" option only.'
+            'Conflicting flags. Unable to specify max-surge/drain-timeout/node-soak-duration with node-image-only.'
+            'If you want to use max-surge/drain-timeout/node-soak-duration with a node image upgrade, please first '
+            'update max-surge/drain-timeout/node-soak-duration using "az aks nodepool update --max-surge/--drain-timeout/--node-soak-duration".'
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/custom.py:1227]: This error message should be the description of the exception on line 1232?

File: src/aks-preview/azext_aks_preview/custom.py [hunk 2]
```diff
@@ -1216,13 +1222,13 @@ def aks_agentpool_upgrade(cmd,
...
         )
 
     # Note: we exclude this option because node image upgrade can't accept nodepool put fields like max surge
-    if max_surge and node_image_only:
+    if (max_surge or drain_timeout or node_soak_duration) and node_image_only::
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/custom.py:1231]: extra trailing colon

**Review by FumingZhang — COMMENTED** [review]
lgtm

Please also add a live test (e2e test) case corresponding to the newly added option, may refer to examples in `test_aks_commands.py`. Don't forget to commit the recording file for the new test case. Get the recording file by running the test case in live mode locally or rerunning the [pipeline](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=81426648&view=results) (set pipeline line variable `EXT_TEST_COVERAGE` to your test case name, if there are multiple, separate them with spaces) once test passed, recording file will be uploaded to pipeline artifact.

File: src/aks-preview/azext_aks_preview/custom.py (hunk 1 above)
Comment (MaxHorstmann) [reply src/aks-preview/azext_aks_preview/custom.py]: Indeed! [Fixed](https://github.com/Azure/azure-cli-extensions/pull/6865/commits/299794bf4d95e280b9acb0cd5ffde679dde968c8)

File: src/aks-preview/azext_aks_preview/custom.py (hunk 2 above)
Comment (MaxHorstmann) [reply src/aks-preview/azext_aks_preview/custom.py]: [fixed](https://github.com/Azure/azure-cli-extensions/pull/6865/commits/ef6e6b602df47afe34d7ab4e6da00497ea104a6f)

File: src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py [hunk 3]
```diff
@@ -1837,6 +1837,45 @@ def test_aks_create_add_nodepool_with_custom_ca_trust_certificates(self, resourc
         self.cmd(
             'aks delete -g {resource_group} -n {name} --yes --no-wait', checks=[self.is_empty()])
 
+    @AllowLargeResponse()
+    @AKSCustomResourceGroupPreparer(random_name_length=17, name_prefix='clitest', location='westus2')
+    def test_aks_nodepool_node_soak_duration(self, resource_group, resource_group_location):
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:1842]: Queued [live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=81489510&view=results) to validate the change.

Comment (MaxHorstmann) [reply src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:1842]: Thanks @FumingZhang! Checking why it failed.

Btw

```
echo 'Live test failed!'
echo 'Please refer to this wiki (https://dev.azure.com/msazure/CloudNativeCompute/_wiki/wikis/CloudNativeCompute.wiki/156735/AZCLI-AKS-Live-Unit-Test-Pipeline) for troubleshooting guidelines.'
```

Looks like that wiki link is outdated. I updated the `WIKI_LINK` pipeline variable [here](https://dev.azure.com/msazure/CloudNativeCompute/_apps/hub/ms.vss-ciworkflow.build-ci-hub?_a=edit-build-definition&id=219428&view=Tab_History), was that the correct url?

Comment (MaxHorstmann) [reply src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:1842]: @FumingZhang so, it looks like the live test is currently failing

> Query 'upgradeSettings.nodeSoakDurationInMinutes' doesn't yield expected value '5', instead the actual value is 'None'

I noticed that your [2023-09-02-preview](https://github.com/Azure/azure-rest-api-specs/tree/FumingZhang-containerservice-Microsoft.ContainerService-2023-09-02-preview) branch (which contains the [Swagger change](https://github.com/Azure/azure-rest-api-specs/pull/26046) for this feature) in Azure/azure-rest-api-specs hasn't been merged yet - I guess that explains it? Let's re-run the test after the Swagger change is merged.

Comment (FumingZhang) [reply src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:1842]: Yes, swagger/sdk has not been updated, so the deserialized response will not include this property.

Comment (FumingZhang) [reply src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:1881]: SDK not updated yet, rebase/merge from main once this PR #6904 got merged

Comment (MaxHorstmann) [reply src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:1881]: Thx - looks like that PR just got merged.

[synced/merged](https://github.com/MaxHorstmann/azure-cli-extensions/commit/065f4a9c44de9381fcbb9128d31da0c2c323cd08) this branch.

queued up new [live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=81902688&view=results)

Comment (FumingZhang) [reply src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:1881]: Seems live test passed, please commit the recording files corresponding to the newly added test cases (find from pipeline artifact).

Please also fix errors in [Static Analysis](https://dev.azure.com/azclitools/public/_build/results?buildId=100166&view=logs&j=ce14e7bd-b9ee-5bd7-1dc4-cb4d70edaf0d&t=90fe1415-141e-56cb-5eb9-d53a9bbb8155&l=3574).

Comment (MaxHorstmann) [reply src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:1881]: > Seems live test passed, please commit the recording files corresponding to the newly added test cases (find from pipeline artifact)

[done](https://github.com/Azure/azure-cli-extensions/pull/6865/commits/1b1eca343377e206efa9b26ae8e34f73ff3a3324)

> Please also fix errors in [Static Analysis](https://dev.azure.com/azclitools/public/_build/results?buildId=100166&view=logs&j=ce14e7bd-b9ee-5bd7-1dc4-cb4d70edaf0d&t=90fe1415-141e-56cb-5eb9-d53a9bbb8155&l=3574).

[done](https://github.com/Azure/azure-cli-extensions/pull/6865/commits/67055f210b96f20ae0d034399e80141aed4452d2)

**Comment by MaxHorstmann** [comment]
@FumingZhang any idea why some of these [checks](https://github.com/Azure/azure-cli-extensions/pull/6865/checks?check_run_id=18123170460) failed? The only error message I see is
```
Bash exited with code '1'.
```

**Comment by FumingZhang** [comment]
> @FumingZhang any idea why some of these [checks](https://github.com/Azure/azure-cli-extensions/pull/6865/checks?check_run_id=18123170460) failed? The only error message I see is
> 
> ```
> Bash exited with code '1'.
> ```

The failure is about test case
> FAILED src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py::AzureKubernetesServiceScenarioTest::test_aks_nodepool_drain_timeout

and seems the recording file corresponding to this case is not committed.

Queue [new live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=82072629&view=results) for this test case.

**Comment by MaxHorstmann** [comment]
> and seems the recording file corresponding to this case is not committed.
> 
> Queue [new live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=82072629&view=results) for this test case.

Thanks @FumingZhang, I added the recording. Let's see if all the checks pass now.

**Review by FumingZhang — APPROVED** [review]
LGTM

**Comment by MaxHorstmann** [comment]
Thanks @FumingZhang! Anything else I need to do to get this merged?

**Comment by yonzhan** [comment]
Please fix conflicting files.

**Comment by MaxHorstmann** [comment]
@yonzhan @FumingZhang 

> Please fix conflicting files.

[done](https://github.com/Azure/azure-cli-extensions/pull/6865/commits/5722956283e81ed29e552a6f014226cded2cd892)

File: src/aks-preview/HISTORY.rst [hunk 4]
```diff
@@ -13,6 +13,11 @@ Pending
 +++++++
 * Vendor new SDK and bump API version to 2023-09-02-preview.
 
+0.5.169
++++++++
+* Add `--node-soak-duration` to the `az aks nodepool add/update/upgrade` commands.
+* Add `--drain-timeout` to the `az aks nodepool add/update/upgrade` commands (already in [azure-cli](https://github.com/Azure/azure-cli/pull/27475)).
```
Comment (zhoxing-ms) [inline src/aks-preview/HISTORY.rst:19]: Same as https://github.com/Azure/azure-cli-extensions/pull/6881#discussion_r1384260534

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**